### PR TITLE
review

### PR DESCRIPTION
**Pull request by omar-calo**

### CODE REVIEW

File: src/services/components/createComponent/handler.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    })
+
+    // Asynchronous Validation
+    try {
+      await validationSchema.validateAsync(event.body);
+    } catch (validationError) {
+      throw createError(400, 'Validation Error', {
+        details: validationError.details.map(detail => detail.message),
+      });
+    }
```
Comment (omar-calo) [inline src/services/components/createComponent/handler.ts:29]: This should be moved to a middleware

File: src/services/components/createComponent/handler.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    const componentData = {
+      ...event.body,
+      name: capitalizeFirstLetter(event.body.name),
```
Comment (omar-calo) [inline src/services/components/createComponent/handler.ts:33]: Is this related to business logic?

File: src/services/components/createComponent/handler.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    };
+    const ingredients = componentData.ingredients;
+    delete componentData.ingredients;
```
Comment (omar-calo) [inline src/services/components/createComponent/handler.ts:36]: Why do we need to do this?

File: src/services/components/createComponent/handler.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    // useCase - Create Component
+    const result = await createComponent(prisma, componentData);
+
+    // useCase - Create ComponentIngredient
+    const parsedResult = JSON.parse(result.body)
+    const componentId = parsedResult.data.id;
+
+    await createComponentIngredient(prisma, componentId, ingredients)
```
Comment (omar-calo) [inline src/services/components/createComponent/handler.ts:45]: Is this business logic 😁 ?

File: src/services/components/createComponent/handler.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    return result;
+  } catch (err) {
+    if (err instanceof Prisma.PrismaClientKnownRequestError && err.code === 'P2002') {
+      console.log('Conflict Error:', err);
+      throw createError(409, 'Conflict Error', {
+        details: err,
+      });
+    }
```
Comment (omar-calo) [inline src/services/components/createComponent/handler.ts:54]: Does the first layer need to know what kind of ORM we're using?

File: src/services/components/createComponent/handler.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    const result = await createComponent(prisma, componentData);
+
+    // useCase - Create ComponentIngredient
+    const parsedResult = JSON.parse(result.body)
+    const componentId = parsedResult.data.id;
+
+    await createComponentIngredient(prisma, componentId, ingredients)
+
+    return result;
+  } catch (err) {
```
Comment (omar-calo) [inline src/services/components/createComponent/handler.ts:48]: Instead of adding `try-catch` to every endpoint, how about creating a middleware, usually exception handler middleware. Easy to implement one

File: src/services/components/createComponent/useCase.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    });
+  }
+}
+
+export async function createComponentIngredient(
+  prisma: PrismaClient,
+  componentId: string,
+  ingredients: Array<Ingredients>
+  ):Promise<any> {
+  const componentRepo = new ComponentRepository(prisma);
```
Comment (omar-calo) [inline src/services/components/createComponent/useCase.ts:42]: For every function, we're creating a new repo instance. How can we make it simpler, but not worrying about doing this everytime?

File: src/services/components/serverless.yml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+createComponent:
+  handler: src/services/components/createComponent/handler.default
+  events:
+    - http:
+        path: v1/component/create
```
Comment (omar-calo) [inline src/services/components/serverless.yml:5]: For create component,  we don't need to add `create` to the path

File: src/services/components/serverless.yml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+searchComponent:
+  handler: src/services/components/searchComponents/handler.default
+  events:
+    - http:
+        path: v1/component/search
+        method: get
+        request:
+          parameters:
+            querystrings:
+              name: true
```
Comment (omar-calo) [inline src/services/components/serverless.yml:24]: Can we use the `getComponents` to do the same thing?

File: src/services/components/getComponents/handler.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import createError from 'http-errors';
+
+const prisma = new PrismaClient();
+
+export default middyfy(async (event) => {
+  try {
+    console.log('Received CloudFormation Event:', JSON.stringify(event, null, 2));
+
+    // UseCase - Get Components
+    const result = await getComponents(prisma);
```
Comment (omar-calo) [inline src/services/components/getComponents/handler.ts:13]: What about if we have 10k components, are we gonna return them all?

Comment (moaz-mostafa23) [reply src/services/components/getComponents/handler.ts:13]: were planning on implementing limiting features to get only a set amount at a time

**Comment by aralhibshi** [comment]
Old review, implemented all
